OverflowException in LoggerMessageGenerator.GetNonRandomizedHashCode

### Description

The source generator in Microsoft.Extensions.Logging.Abstractions throws OverflowException when it attempts to generate an EventId for some logger method names.  Pronounceable names that trigger this error include `Dzoggee`, `Juzzwenn`, `Bzuengdom`, `Covieeszu`, `Ennudkilw`, `Fawiinska`, `Haikshyke`, `Hyklyytja`, `Kirraphas`, `Nuthvausu`, `Orluavjee`, `Piaanwatz`, `Ponkpauij`, `Ruakszyrn`, and `Sazkwinuo`.

### Reproduction Steps

EventIdHashOverflow.csproj:

```XML
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>net8.0</TargetFramework>
    <Nullable>enable</Nullable>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Microsoft.Extensions.Logging.Abstractions" Version="8.0.2" />
  </ItemGroup>

</Project>
```

LoggerExtensions.cs:

```csharp
using Microsoft.Extensions.Logging;

namespace EventIdHashOverflow
{
    internal static partial class LoggerExtensions
    {
        [LoggerMessage(Level = LogLevel.Information)]
        public static partial void Dzoggee(this ILogger logger);
    }
}
```

Attempt to build this project.

### Expected behavior

The build should succeed.

### Actual behavior

CSC : warning CS8785: Generator 'LoggerMessageGenerator' failed to generate source. It will not contribute to the output and compilation errors may occur as a result. Exception was of type 'OverflowException' with message 'Negating the minimum value of a twos complement number is invalid.'.

LoggerExtensions.cs(8,36,8,43): error CS8795: Partial method 'LoggerExtensions.Dzoggee(ILogger)' must have an implementation part because it has accessibility modifiers.


### Regression?

_No response_

### Known Workarounds

Specify LoggerMessageAttribute.EventId instead of letting it be generated automatically.

### Configuration

.NET SDK 9.0.101

### Other information

OverflowException is thrown by the Math.Abs(Int32.MinValue) call here: <https://github.com/dotnet/runtime/blob/9cb3b725e3ad2b57ddc9fb2dd48d2d170563a8f5/src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Parser.cs#L850-L862>

There is similar code in other source generators too:

* <https://github.com/dotnet/runtime/blob/5a9362f509629304c91ae07662f649282f956552/src/libraries/Microsoft.Extensions.Options/gen/OptionsSourceGenContext.cs#L68-L81>
* <https://github.com/dotnet/extensions/blob/16acdbf1442f579862cfea15ffd0782510078ff0/src/Generators/Microsoft.Gen.Logging/Emission/Emitter.Method.cs#L141-L159>

Making GetNonRandomizedHashCode return 0 in this case would make every nonnegative `int` return value correspond to two values of `uint result`.

--- a/src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Parser.cs
+++ b/src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Parser.cs
@@ -914,7 +914,7 @@ namespace Microsoft.Extensions.Logging.Generators
             uint result = 2166136261u;
             foreach (char c in s)
             {
-                result = unchecked((c ^ result) * 16777619);
+                result = (c ^ result) * 16777619;
             }
 
             return (int)(result & 0x7FFFFFFF); // Ensure the result is non-negative

--- a/src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Parser.cs
+++ b/src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Parser.cs
@@ -914,9 +914,10 @@ namespace Microsoft.Extensions.Logging.Generators
             uint result = 2166136261u;
             foreach (char c in s)
             {
-                result = (c ^ result) * 16777619;
+                result = unchecked((c ^ result) * 16777619);
             }
-            return Math.Abs((int)result);
+
+            return (int)(result & 0x7FFFFFFF); // Ensure the result is non-negative
         }
     }
 }

--- a/src/libraries/Microsoft.Extensions.Logging.Abstractions/tests/Microsoft.Extensions.Logging.Generators.Tests/LoggerMessageGeneratorParserTests.cs
+++ b/src/libraries/Microsoft.Extensions.Logging.Abstractions/tests/Microsoft.Extensions.Logging.Generators.Tests/LoggerMessageGeneratorParserTests.cs
@@ -535,7 +535,7 @@ namespace Microsoft.Extensions.Logging.Generators.Tests
                     public partial void M1();
                 }
             ");
-            
+
             Assert.Equal(2, diagnostics.Count);
 
             Assert.Equal(DiagnosticDescriptors.PrimaryConstructorParameterLoggerHidden.Id, diagnostics[0].Id);
@@ -986,6 +986,23 @@ namespace Microsoft.Extensions.Logging.Generators.Tests
             ");
 
             Assert.Empty(diagnostics);    // should fail quietly on broken code
+        }
+
+        [Fact]
+        public async Task EventIdGenerationTest()
+        {
+            // hashing Dzoggee should not result breaking generation.
+            // in our hashing, Dzoggee result to int.MinValue before we convert it to non-negative value.
+            IReadOnlyList<Diagnostic> diagnostics = await RunGenerator(@"
+                using Microsoft.Extensions.Logging;
+                internal static partial class LoggerExtensions
+                {
+                    [LoggerMessage(Level = LogLevel.Information)]
+                    public static partial void Dzoggee(this ILogger logger);
+                }
+            ");
+
+            Assert.Empty(diagnostics);
         }
 
         [Fact]

--- a/src/libraries/Microsoft.Extensions.Options/gen/OptionsSourceGenContext.cs
+++ b/src/libraries/Microsoft.Extensions.Options/gen/OptionsSourceGenContext.cs
@@ -67,17 +67,16 @@ namespace Microsoft.Extensions.Options.Generators
 
         /// <summary>
         /// Returns a non-randomized hash code for the given string.
-        /// We always return a positive value.
         /// </summary>
-        internal static int GetNonRandomizedHashCode(string s)
+        internal static uint GetNonRandomizedHashCode(string s)
         {
             uint result = 2166136261u;
             foreach (char c in s)
             {
-                result = (c ^ result) * 16777619;
+                result = unchecked((c ^ result) * 16777619);
             }
 
-            return Math.Abs((int)result);
+            return result;
         }
     }
 }

--- a/src/libraries/Microsoft.Extensions.Options/gen/OptionsSourceGenContext.cs
+++ b/src/libraries/Microsoft.Extensions.Options/gen/OptionsSourceGenContext.cs
@@ -73,7 +73,7 @@ namespace Microsoft.Extensions.Options.Generators
             uint result = 2166136261u;
             foreach (char c in s)
             {
-                result = unchecked((c ^ result) * 16777619);
+                result = (c ^ result) * 16777619;
             }
 
             return result;

